src/scripts/ui/inline-roll-links.ts

```diff
@@ -294,39 +293,39 @@ export const InlineRollLinks = {
                     }
                 }
 
-                const flags: { pf2e: Record<string, unknown> } = {
-                    pf2e: {},
-                };
+                const pf2e = templateData.flags?.pf2e ?? {};
 
                 if (
                     objectHasKey(CONFIG.PF2E.areaTypes, pf2EffectArea) &&
                     objectHasKey(CONFIG.PF2E.areaSizes, templateData.distance)
                 ) {
-                    flags.pf2e.areaType = pf2EffectArea;
+                    pf2e.areaType = pf2EffectArea;
                 }
 
                 const messageId =
                     foundryDoc instanceof ChatMessagePF2e
                         ? foundryDoc.id
                         : htmlClosest(html, "[data-message-id]")?.dataset.messageId ?? null;
                 if (messageId) {
-                    flags.pf2e.messageId = messageId;
+                    pf2e.messageId = messageId;
                 }
 
                 const actor = resolveActor(foundryDoc, link);
-                if (actor || pf2Traits) {
-                    const origin: Record<string, unknown> = {};
-                    if (actor) {
-                        origin.actor = actor.uuid;
-                    }
-                    if (pf2Traits) {
-                        origin.traits = pf2Traits.split(",");
-                    }
-                    flags.pf2e.origin = origin;
+                const origin = isRecord(pf2e.origin) ? pf2e.origin : {};
```

Comment:
@stwlam  I changed the code to use a type guard instead.  Is this good enough?  I could not find another way other way other than changing the documents types so that pf2e is Record<string, Record<string, unknown>> over Record<string, unknown>. That or using the any value from a json serialization.